SE-NNNN: Build-Time Constant Values

artemcm opened the pull request.

artemcm committed:
SE-0353: Build-Time Constant Values

airspeedswift commented on proposals/XXXX-build-time-constant-values.md:
```diff
@@ -0,0 +1,188 @@
+
+## Introduction
+
+A Swift language feature for designating certain values to be knowable at compile-time consisting of an attribute to designate certain properties, and function parameters to have compile-time knowable values. Such information can be used by the compiler to extract build-time program information and to serve as a foundation for richer compile-time features in the future.
```
```suggestion
A Swift language feature for requiring certain values to be knowable at compile-time. This is achieved through an attribute, `@const`, constraining properties and function parameters to have compile-time knowable values. Such information forms a foundation for richer compile-time features in the future, such as extraction and validation of values at compile time.
```

airspeedswift commented on proposals/XXXX-build-time-constant-values.md:
```diff
@@ -0,0 +1,188 @@
+
+## Proposed Solution
+
+The Swift compiler will recognize declarations of properties, local variables, and function parameters declared with a `@const` attribute as having an additional guarantee to be known at compile-time. If a `@const` property or variable is initialized with a runtime value, the compiler will emit an error. Similarly, if a runtime value is passed as an argument to a `@const` function parameter, the  ...
```
```suggestion
The Swift compiler will recognize declarations of properties, local variables, and function parameters declared with a `@const` attribute as having an additional requirement to be known at compile-time. If a `@const` property or variable is initialized with a runtime value, the compiler will emit an error. Similarly, if a runtime value is passed as an argument to a `@const` function parameter, the compiler will emit an error. Aside from participating in name mangling, the attribute has no runtime effect.
```

airspeedswift commented on proposals/XXXX-build-time-constant-values.md:
```diff
@@ -0,0 +1,188 @@
+* Array literals consisting of integer, floating-point, string literals or enum cases with no associated values.
+* Dictionary literals consisting of integer, floating-point, string literals or enum cases with no associated values.
+
+This list may expand in the future to include more literal-value kinds or potential new compile-time valued constructs.
```
Can probably say "will" expand. Could even give an example of @frozen structs with public members.

airspeedswift commented on proposals/XXXX-build-time-constant-values.md:
```diff
@@ -0,0 +1,188 @@
+* Integer and floating-point literals.
+* String literals (excluding interpolated strings).
+* Enum cases with no associated values.
+* Array literals consisting of integer, floating-point, string literals or enum cases with no associated values.
```
can probably just say "Array and Dictionary literals consisting of elements in this list" (do we support nested arrays?)

airspeedswift commented on proposals/XXXX-build-time-constant-values.md:
```diff
@@ -0,0 +1,188 @@
+* Compile-time types - allow types consisting of `const` properties to be treated as compile-time-known values.
+* Compile-time expressions - allow expressions that operate on compile-time-known values, from binary operators to control flow.
+* Compile-time functions - consisting of expressions operating on compile-time-known values and `const` parameters.
+
```
I would add "toolchain support for extracting compile-time values at compile time". Today you can do it using code inspection tools like SourceKit but in future a more structured support for this kind of extraction could be built.

airspeedswift replied:
Also we should mention propagation of constants.

airspeedswift reviewed (commented):
I think the main thing still needed is Alternatives Considered.

Two areas to cover here:
- why an attribute not a keyword or an introducer
- isn't this just the same as `let`? (partly answered by the URL initializer)

airspeedswift commented on proposals/XXXX-build-time-constant-values.md:
```diff
@@ -0,0 +1,188 @@
+
+## Motivating Example Use-Cases
+
+### Declarative Package Manifests that do not require execution
```
This section heading differs from the other because it cites the example, not the general design goal. Suggest renaming it to "Facilitate Compile-time Extraction of Values".

alexbbrown committed:
Update 0341-opaque-parameters.md (#1615)

very tiny typo

rxwei committed:
Amend SE-0351 and SE-0350 (#1612)

SE-0351:
- Add future direction about conversion to textual regex.
- Move recursive subpatterns to future directions.
- Clarify `Regex.Match.subscript(_:)` precondition.

SE-0350:
- Rename `firstMatch(_: Substring)` to `firstMatch(in: Substring)` to be consistent with the `String` variant.
- Specify accessors on properties to clarify mutability.

glessard committed:
typo fix

glessard committed:
Merge pull request #1617 from glessard/unaligned-loads

Capitalization fix in SE-0349

lorentey committed:
Amend SE-0329 to add Clock.Duration

airspeedswift committed:
Update 0350-regex-type-overview.md

airspeedswift committed:
Update 0351-regex-builder.md

rjmccall committed:
Merge pull request #1618 from lorentey/clock-duration

Amend SE-0329 to add Clock.Duration

rjmccall committed:
Accept SE-0346

rjmccall committed:
Merge pull request #1620 from apple/accept-se-0346

Accept SE-0346

rjmccall committed:
Accept SE-0349

rjmccall committed:
Merge pull request #1621 from apple/accept-se-0349

Accept SE-0349

artemcm committed:
Fix placement of '@const' attribute on parameters type

artemcm committed:
Fixup URL forum thread URL

artemcm committed:
Fix Typos

artemcm committed:
Qualify supported types: add booleans and tuples

artemcm committed:
Add discssion on propagation rules, memory placement and type annotation

amartini51 commented:
Reading through this draft to scope documentation impact and start drafting.  Not sure why this PR includes a bunch of minor changes to other proposals too — happy to help with the Git commands to separate that out if it gives you trouble.

artemcm commented:
> Reading through this draft to scope documentation impact and start drafting. Not sure why this PR includes a bunch of minor changes to other proposals too — happy to help with the Git commands to separate that out if it gives you trouble.

Not sure what happened there, but the latest version is a PR against the main repo here:
https://github.com/apple/swift-evolution/pull/1622
